lib/internal/modules/esm/loader.js

```diff
@@ -45,6 +46,16 @@ class Loader {
     // an object with the same keys as `exports`, whose values are get/set
     // functions for the actual exported values.
     this._dynamicInstantiate = undefined;
+
+    // Set up context passed to hooks
+    const k = Object.freeze(Object.create(null));
+    vmModuleLinkHookMap.set(k, this);
+
+    this.hookContext = Object.assign(Object.create(null), {
+      defaultResolve,
+      resolve: (specifier, parentURL) => this.resolve(specifier, parentURL),
+      vmModuleLinkHook: k,
+    });
```

Comment:
If you want to deny the hook context access to outside world, you are not – the [[Prototype]] of these functions are the %FunctionPrototype% of the outside world. It would be better to create these functions in the context directly, if possible, rather than using `Object`/`Reflect.setPrototypeOf()`.